v2/pkg/protocols/common/protocolstate/headless.go

```diff
@@ -0,0 +1,77 @@
+package protocolstate
+
+import (
+	"strings"
+
+	"github.com/go-rod/rod"
+	"github.com/go-rod/rod/lib/proto"
+	"github.com/projectdiscovery/networkpolicy"
+	errorutil "github.com/projectdiscovery/utils/errors"
+	urlutil "github.com/projectdiscovery/utils/url"
+	"go.uber.org/multierr"
+)
+
+// initalize state of headless protocol
+
+var (
+	ErrURLDenied         = errorutil.NewWithFmt("headless: url %v dropped by rule: %v")
+	networkPolicy        *networkpolicy.NetworkPolicy
+	allowLocalFileAccess bool
+)
+
+// ValidateNFailRequest validates and fails request
+// if the request does not respect the rules, it will be canceled with reason
+func ValidateNFailRequest(page *rod.Page, e *proto.FetchRequestPaused) error {
+	reqURL := e.Request.URL
+	normalized := strings.ToLower(reqURL) // normalize url to lowercase
```

Comment:
Internally google normalize urls like `                            http://www.xxx.yy` to `http://www.xxx.yy`. I'm not sure if at `FetchRequestPaused` event this normalization already happen (most likely yes as the request is already compiled), anyway just in case maybe we can also trim spaces and other format directives such as `\t\n` (probably they are not even allowed).